cvmfs/network/download.cc

```diff
@@ -1797,13 +1702,14 @@ Failures DownloadManager::Fetch(JobInfo *info) {
     LogCvmfs(kLogDownload, kLogDebug, "download failed (error %d - %s)", result,
              Code2Ascii(result));
 
-    if (info->destination == kDestinationPath)
-      unlink(info->destination_path->c_str());
+    if (info->destination == kDestinationPath) {
+      unlink(static_cast<cvmfs::PathSink*>(info->destination_sink)->
+                                                                 path_.c_str());
+    }
 
-    if (info->destination_mem.data) {
-      free(info->destination_mem.data);
-      info->destination_mem.data = NULL;
-      info->destination_mem.size = 0;
+    if (info->destination == kDestinationMem) {
+      // This IF could be removed? just always reset the sink?
+      info->destination_sink->Reset();
```

Comment:
Let's handle this consistently with a virtual `Sink::Purge()` method that releases all the resources of the sink.  I.e., memory sink frees the memory, path sink unlinks the file, file sink closes the fd. The `Purge()` method would only take action if the sink is owning and then leave the sink in an invalid state.